plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -53,123 +69,133 @@ in
+
+-- | Apply a list of term and type arguments to a function in potentially a nested fashion.
+mkApps :: Term tyname name uni fun ann
+  -> ArgsWithAnns tyname name uni fun ann
+  -> Term tyname name uni fun ann
+mkApps f ((MkTermArg tmArg, ann) : args) = mkApps (Apply ann f tmArg) args
+mkApps f ((MkTypeArg tyArg, ann) : args) = mkApps (TyInst ann f tyArg) args
+mkApps f []                              = f
+
+-- | Given the arity of a function, and the list of arguments applied to it, return whether it is
+-- fully applied or not.
+isFullyApplied :: Arity -> Args tyname name uni fun ann -> Bool
+isFullyApplied [] _ = True -- The function needs no more arguments
+isFullyApplied (_lam:_lams) [] = False -- under-application
+isFullyApplied (hdLams:tlLams) (hdArg:tlArg) =
+  case (hdLams, hdArg) of
+    (TermParam, MkTermArg _) -> isFullyApplied tlLams tlArg
+    (TypeParam, MkTypeArg _) -> isFullyApplied tlLams tlArg
+    _                        ->
+      -- wrong argument type, i.e., we have an ill-typed term here. It's not what we define as fully
+      -- applied. Although if the term was ill-typed before, it will be ill-typed after the
+      -- inlining, and it won't make it any worse, so we could consider accepting this.
+      False
+
+-- | Consider whether to inline a single node. We inline fully applied functions iff the body of
+-- the function is `acceptable`.
+considerInlineSat :: forall tyname name uni fun ann. InliningConstraints tyname name uni fun
+    => Term tyname name uni fun ann -- ^ The `body` of the `Let` term.
+    -> InlineM tyname name uni fun ann (Term tyname name uni fun ann)
+considerInlineSat tm = do
+    -- collect all the arguments of the term being applied to
+    let (fun, args) = collectArgs tm
+    case fun of
+      -- if it is a `Var` that is being applied to, check to see if it's fully applied
+      v@(Var _ann name) -> do
+        maybeVarInfo <- gets (lookupVarInfo name)
+        case maybeVarInfo of
+          -- the variable maybe a *recursive* let binding, in which case it won't be in the map,
+          -- and we don't process it. ATM recursive bindings aren't inlined.
+          Nothing -> pure tm
+          Just varInfo -> do
+            let isAcceptable = acceptable (varBody varInfo)
+            -- if the body of `Var` is `acceptable` and
+            -- it is fully applied (over-application is allowed) then
+            if isAcceptable && isFullyApplied (arity varInfo) (map fst args)
+            then do
+              -- check if term is pure. Because non-functions or binding with arity 0 are always
+              -- fully applied, we only want to inline them if they are also pure.
+              -- For strict bindings we can't accidentally make any effects happen less often than
+              -- it would have before, but we can make it happen more often.
+              -- We could potentially do this safely in non-conservative mode.
+              isTermPure <- checkPurity v
```

Comment:
This is checking the wrong thing? We want to check if the RHS of the binding we're about to inline is pure, i.e. `varDef varInfo`.

Reply:
I'm confused. Yes, we check purity as a proxy for duplicating effects often, but there's nothing wrong with doing it here?

Reply:
I'm just saying that I'll say that I'm checking `wontDuplicateEffect` here, which is what I really want, which includes checking purity AND strictness. Am I making sense? (same as the comment below)

Reply:
OK I just saw your other comment, so I'll do `isTermBindingPure` here as well.